PIP-25: Token based authentication

merlimat opened the pull request:
### Motivation

This is the first PR for the implementation of token based authentication

### Modifications

 * Added client & broker auth providers
 * Added CLI tool to manage tokens
 * Unit tests

TODOs (in separate PRs): 
 * Integration tests
 * Documentation
 * C++ (and related) client plugin implementation

maskit commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:
```diff
@@ -0,0 +1,116 @@
+@Slf4j
+public class AuthenticationProviderToken implements AuthenticationProvider {
+
+    public final static String HTTP_HEADER_NAME = "X-Pulsar-Auth";
```
Why not `Authorization` header?
https://tools.ietf.org/html/rfc6750#section-2.1

merlimat replied:
Good point. I have fixed to use `Authorization`

ivankelly commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:
```diff
@@ -0,0 +1,116 @@
+
+    public final static String HTTP_HEADER_NAME = "X-Pulsar-Auth";
+
+    final static String CONF_TOKEN_SECRET_KEY = "tokenSecretKey";
```
Ok, it looks like setSigningKey() below can take either a PublicKey or a SecretKey, so the variable naming here is all off. 

So these conf strings should change to reflect this. s/Secret/Signing/g

merlimat replied:
I have split the config variables in 2 sets: one for `secretKey` the other for `publicKey`, to avoid such confusion. I'll add documentation later on how to configure a cluster with token authentication and I'll specify the config that applies in each case.

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    public static long parseRelativeTimeInMillis(String relativeTime) {
```
Yes, I have refactored into `RelativeTimeUtil` in `pulsar-common`

merlimat commented on pom.xml:
```diff
@@ -726,6 +727,22 @@ flexible messaging model and an intuitive client API.</description>
       </dependency>
+      
+      <dependency>
+        <groupId>io.jsonwebtoken</groupId>
```
Fixed. The jackson dependency is not a big problem because we pin the version we're importing

merlimat commented:
@ivankelly @maskit Please take another look

merlimat commented:
> Also, it would be really good to have an integration test in this PR a) to validate that both the PSK and private/public paths work 

The 2 paths are tested in the auth provider unit test. It's not the same as integration test but it covers all the plugin code, from getting variables from `ServerConfiguration` to validating the tokens in both cases.

maskit commented on conf/broker.conf:
```diff
@@ -274,6 +274,24 @@ athenzDomainNames=
+# Read secret key from an environment variable
+tokenSecretKeyFromEnv=
+# Read secret key from a file
+tokenSecretKeyFromFile=
```
Are there any case that we want to read tokenSecretKey from both Env and File?
How about using URL scheme?
```
tokenSecretKey=env:///FOO_BAR
tokenSecretKey=file:///path/for/file
```

Once we added support for "env:///" (not sure if we need 3 slashes) to `org.apache.pulsar.client.api.url.URL`, every places that use the class would support reading configuration from environment variable.

Reference: #672, #1212

ivankelly replied:
not really related, but Tim Berners Lee has said the // in http is a mistake. I guess file took the double slash to ape http, the third is to make the path absolute. Env should have at most 2 slashes, but none are needed.

ivankelly replied:
Also, it becomes a URI not a URL since it's not really pointing to a location

merlimat replied:
Good point. I'll merge the 3 keys into a single one and make sure the same conf also applies to `client.conf` for consistency.

maskit commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,222 @@
+                "--stdin" }, description = "Read secret key from standard input")
+        private Boolean stdin = false;
+
+        @Parameter(names = { "-pk" }, description = "Indicate the key is ")
```
the key is ... a private key?

maskit commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:
```diff
@@ -0,0 +1,147 @@
+
+        if (validationKeyConfig.startsWith("data:")) {
+            log.info("Reading token validation key from config file");
+            validationKey = validationKeyConfig.substring("data:".length());
```
This might be valid for a super simple case, but it wouldn’t work for cases that has mime-type or base64 encoding.

Why didn’t you use the URL class?

merlimat replied:
... because I've never used `URL.openStream()` before :) 

Though, one thing is that the key is expected to be already base64 encoded. 

```shell
$ bin/pulsar tokens create-secret-key
secret-key CBqtuReZe+8nKqtJBbmuPTIg6b4W1KjOS21pcJkTJ/8=
```

Then one would just have to store the key string somewhere.

maskit replied:
I’m ok with parsing the schemes here and having a limitation in its format for now, but I still think this data scheme parsing is incorrect. It always need a comma after the colon. 
https://tools.ietf.org/html/rfc2397#section-3

Also, base64 encoding is not mandatory. You can pass URL encoded string instead.

merlimat replied:
@maskit I have updated to conform with `data:` standard url, though I had to add manual parsing since java `URL` class doesn't support it.

maskit replied:
Did you look into #1212 I commented on another line? The URL class I suggested is `org.apache.pulsar.client.api.url.URL`. The parser already exists in it.

merlimat commented:
@maskit @ivankelly please take another look. I have moved to use standard `data:` uri and to not force base64 encodings.

maskit commented:
The "data:" parser you wrote would be a duplicated code because it's already implemented in in-house URL class I wrote.  I'm not going to make a change request because I don't want to delay this feature, but either of the parsers should be removed eventually to eliminate duplicate code.

Also, because environment variables are general enough as data sources, I think "env:" parser should be added as a `URLStreamHandler` so that other places can use it too via the in-house URL class.

As for "token:", I'm ok with having it in this plugin as a special case. But if you renamed it to "raw:", it could be used on other places.

maskit commented:
If we added all the parsers as URLStreamHandlers, users would be able to read any of them like this without taking care of the scheme.
```java
import org.apache.pulsar.client.api.url.URL;
import java.net.URLConnection;

URLConnection urlConnection = new URL(keyConfUrl).openConnection();
String keyData = CharStreams.toString(new InputStreamReader((InputStream) urlConnection.getContent()));
```

Since `openConnection()` is defined in `java.net.URL`. The only thing users need to do is changing `import` line. Then they can use it as the same as `java.net.URL`.
https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#openConnection--

I don't remember why I didn't implement `URL#getContent()` (it's a shorthand for `openConnection().getContent()`) but if we did, it would be much easier.

merlimat commented:
@maskit I got now what you meant. I had forgot about `org.apache.pulsar.client.api.url.URL`. There is just one problem with that class in that it assumes the data will be decoded into a `String`, though in this cases it would need to be a `byte[]` since the key cannot be converted into `String`.

Will continue this in a different PR since this will need to change on Athenz plugin too.

merlimat commented:
run integration tests

ivankelly commented on conf/broker.conf:
```diff
@@ -283,6 +283,24 @@ athenzDomainNames=
 
+### --- Token Authentication Provider --- ###
+
+## Simmetric key
```
Symmetric

merlimat replied:
Done

ivankelly commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/utils/AuthTokenUtils.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    private static byte[] readDataUrl(String data) throws IOException {
+        // Expected format is:
```
why is this more complex than data:<base64str>? It seems like it's just making it more complex to document and test.

maskit replied:
What do you mean? This is the format of data:.
https://tools.ietf.org/html/rfc2397#section-2

merlimat replied:
Changed it to use existing `URL` class

ivankelly replied:
@maskit ah, i wasn't aware it was conforming to a standard.

ivankelly commented on pulsar-broker-common/src/test/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenTest.java:
```diff
@@ -0,0 +1,287 @@
+        provider.close();
+    }
+
+    @Test
```
doesn't look like you have a test for env:

ivankelly replied:
I was sure that System.setenv existed. Seems not.

ivankelly commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,231 @@
+        String outputFile;
+
+        @Parameter(names = {
+                "-b", "--base-64" }, description = "Encode the key in base64")
```
```--base64```

ivankelly replied:
Shouldn't this be enabled by default in the case that -o is not specified?

ivankelly replied:
still, ```--base64```

ivankelly commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,231 @@
+        private Boolean isPrivateKey = false;
+
+        @Parameter(names = { "-k",
+                "--signing-key" }, description = "Pass the signing key. This can either be: data:, file:, etc..", required = true)
```
```--signing-key``` is a new concept when you compare it the "create" commands. it would be better to have
```--private-key``` and ```--secret-key```, require that only one of them is specified, and kill the ```--is-private-key``` flag.

ivankelly replied:
Especially since there's no concept of signing key in the broker conf.

merlimat replied:
Good point, fixed

ivankelly commented on conf/broker.conf:
```diff
@@ -283,6 +283,24 @@ athenzDomainNames=
 # When this parameter is not empty, unauthenticated users perform as anonymousUserRole
 anonymousUserRole=
 
+### --- Token Authentication Provider --- ###
```
need same in proxy.conf also.

ivankelly replied:
Also, site2/docs/reference-configuration.md (for both broker and proxy)

merlimat replied:
Fixed. Though, we'd really should automate this. It's already wildly out of date.

ivankelly replied:
+1000 for automation

ivankelly commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,231 @@
+                token = r.readLine();
+            } else if (tokenFile != null) {
+                token = new String(Files.readAllBytes(Paths.get(tokenFile)), Charsets.UTF_8);
+            } else if (System.getenv("TOKEN") != null) {
```
why is this here? Seems unnecessary. Anyone who wanted to use this would just pass $TOKEN as the parameter on the shell

merlimat replied:
The main reason is that a "token" if effectively like a password (or a private key): whoever has it, can access the system. Passing that as  a CLI argument will expose to other users on the same machine.

ivankelly replied:
and how would they get it into the environment?

ivankelly replied:
In any case, I'm fine with it staying, just not sure how much real security it offers and trying to reduce options.

merlimat replied:
> and how would they get it into the environment?

I guess you can do: 
```
TOKEN=xyz pulsar tokens show
```

The env variable will not be visible on the process list

ivankelly commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,231 @@
+            String token;
+            if (args != null) {
+                token = args.get(0);
+            } else if (stdin) {
```
It's pretty standard in unix for '-' to denote stdin. 
So: ```token show -f -```

merlimat replied:
I thought to make it explicit for easier discoverability

ivankelly replied:
sure

ivankelly commented on pulsar-broker/src/main/java/org/apache/pulsar/utils/auth/tokens/TokensCliUtils.java:
```diff
@@ -0,0 +1,231 @@
+                token = System.getenv("TOKEN");
+            } else {
+                System.err.println(
+                        "Token needs to be either passed through `--stdin`, `--token-file` or by `TOKEN` environment variable");
```
omits the fact that a argument to the command works.

ivankelly commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationTls.java:
```diff
@@ -41,6 +41,14 @@
     private String certFilePath;
     private String keyFilePath;
 
+    public AuthenticationTls() {
```
Seems unrelated

ivankelly replied:
ok

ivankelly commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java:
```diff
@@ -0,0 +1,104 @@
+    public void configure(String encodedAuthParamString) {
+        // Interpret the whole param string as the token. If the string contains the notation `token:xxxxx` then strip
+        // the prefix
+        if (encodedAuthParamString.startsWith("token:")) {
```
how is this different to data: on the server side?

merlimat replied:
Data is meant to carry arbitrary binary data. A token is guaranteed to be a valid ASCII string (being composed of base64 encoded parts). 

The `encodedAuthParamString` are common to all the auth providers and are a way to pass implementation specific configs. In case of token auth, we're taking the "token" string itself.

I didn't want to use `data:` here to simplify the user input.

ivankelly replied:
ok

maskit replied:
It should be all or nothing, no? I don't see reasons to support token, file and env but not data.

ivankelly replied:
no, token instead of data makes sense. token: is used for passing the token to the client. data: is used to allow the server to load the signing key. They're different usecases. The signing key could be quite large in case of public-private key, whereas the token is generally quite small.

maskit replied:
Why we shouldn't support data: for small data? Even if it's not a suitable scheme, I don't see reasons to drop it. If we allowed using any schemes for any cases and used the URL class to support them, we wouldn't need to maintain this if-else block.

maskit commented:
@merlimat 

> There is just one problem with that class in that it assumes the data will be decoded into a String, 

getContent() returns a InputStream so you should be able to read raw data.

> Will continue this in a different PR since this will need to change on Athenz plugin too.

Sure, I think I can help you out on this since Pulsar seems like compilable with Java 11 :)

ivankelly commented on tests/integration/src/test/java/org/apache/pulsar/tests/integration/auth/token/PulsarTokenAuthenticationBaseSuite.java:
```diff
@@ -0,0 +1,178 @@
+
+        @Cleanup
+        PulsarClient client = PulsarClient.builder()
+                .serviceUrl(pulsarCluster.getPlainTextServiceUrl())
```
not for this PR, but do we have a requireTLS option? We should have an option to block authorization unless TLS is in use to avoid people sending tokens over cleartext.

merlimat replied:
I think there was another PR that was meant to allow to disable the clear-text service endpoints.

ivankelly commented on tests/integration/src/test/java/org/apache/pulsar/tests/integration/auth/token/TokenAuthWithPublicPrivateKeys.java:
```diff
@@ -0,0 +1,76 @@
+@Slf4j
+public class TokenAuthWithPublicPrivateKeys extends PulsarTokenAuthenticationBaseSuite {
+
+    private static final String PRIVATE_KEY_PATH = "/tmp/private.key";
```
nit: non-portable. use File#createTempFile() instead

merlimat replied:
The file is created inside the container and then copied out. I can change the outside file only

ivankelly replied:
Ah, I missed that this was container based. Ignore the comment then.

ivankelly commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/utils/AuthTokenUtils.java:
```diff
@@ -90,39 +92,18 @@ public static String createToken(Key signingKey, String subject, Optional<Date>
+            } catch (Exception e) {
+                throw new IOException(e);
+            }
         } else if (keyConfUrl.startsWith("env:")) {
```
Couldn't you have a URL handler for this also?

merlimat replied:
Yes, adding

merlimat replied:
Actually, there's one tricky part. 

The key, passed as an env variable, needs to be base64 encoded. That might not be needed in general case for `env:` (eg: a token would be fine as it is). 

It's kind of confusing and we'd have to define something similar to `env:base64,MY_SECRET_KEY_VAR`

maskit replied:
I think we shoudn’t define encoding field on env scheme to keep it simple. The requirement comes from application use case (which is this plugin here), so the application should take care of it.
If you get URLConnection from the URL object, you can check the scheme name as protocol name.

maskit replied:
I didn’t read key generation part, but if the key is in PEM format we can use the same format for all scheme and decode the raw data read from the InputStream in a consistent way, I think.

merlimat replied:
I have removed the `env:` schema as it was not working properly. Java seems to be mangling long var values. We can add later when we can make it work.

ivankelly commented:
When you do the integration tests, we need a test for proxy (it seems to be missing in this and c++ change)

maskit commented on pulsar-common/src/main/java/org/apache/pulsar/client/api/url/DataURLStreamHandler.java:
```diff
@@ -57,20 +60,23 @@ public void connect() throws IOException {
                 if (matcher.group("base64") == null) {
                     // Support Urlencode but not decode here because already decoded by URI class.
-                    this.data = new String(matcher.group("data").getBytes(), charset);
+                    this.data = matcher.group("data").getBytes();
```
Well, this is probably equivalent to the original code, so this change is fine.

However, I realized that the both code have the same issue. Because we drop charset information here, users cannot read the data as a string with the specified charset. I think we should stop parsing parameters in mimetype and return the whole mimetype part as content-type.

merlimat replied:
The problem with original code is that it was getting a list of bytes and converting them into a `String` (which in case of keys is not possible) and then exposing back to a `ByteArrayInputStream`. I just cut through the String conversion

maskit replied:
Yeah, I can see that, and we shoud fix that part. So this would be out of this PR’s scope but we need to return charset as a part of contentType so that URL class users can convert the data into String after reading the raw data from InputStream. I think we can do this by removing charset part from the regex.

maskit commented on pulsar-common/src/main/java/org/apache/pulsar/client/api/url/DataURLStreamHandler.java:
```diff
@@ -57,20 +60,23 @@ public void connect() throws IOException {
                 }
+
+                for (int i =0; i < matcher.groupCount(); i++) {
+                    System.out.println("Group: " + matcher.group(i));
```
Do we want to keep this?

merlimat replied:
Ups, forgot when debugging

merlimat commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/utils/AuthTokenUtils.java:
```diff
@@ -0,0 +1,109 @@
+        } else if (keyConfUrl.startsWith("env:")) {
+            String envVarName = keyConfUrl.substring("env:".length());
+            return Decoders.BASE64.decode(System.getenv(envVarName));
+        } else {
```
There's no token used on server side. It would only be either secret, private or public key.

maskit replied:
got it.

merlimat commented:
> When you do the integration tests, we need a test for proxy (it seems to be missing in this and c++ change)

@ivankelly the test is actually going trough the proxy already since `pulsarCluster.getPlainTextServiceUrl()` returns the proxy container address

merlimat commented:
@maskit @ivankelly Few more changes:
 * Fixed integration tests (test is also going against proxy)
 * Added CLI tools to validate a token against a key
 * Removed `env: ` as it's not working properly.

maskit commented on pulsar-common/src/main/java/org/apache/pulsar/client/api/url/DataURLStreamHandler.java:
```diff
@@ -57,20 +60,23 @@ public void connect() throws IOException {
-                if (charset == null) {
-                    charset = "US-ASCII";
+                if (contentType == null) {
+                    this.contentType = "application/data";
```
```suggestion
                    this.contentType = "text/plain;charset=US-ASCII";
```

ivankelly commented on tests/integration/src/test/java/org/apache/pulsar/tests/integration/utils/DockerUtils.java:
```diff
@@ -236,6 +239,82 @@ public void onComplete() {
+        String cmdString = Arrays.stream(cmd).collect(Collectors.joining(" "));
+        ByteBuf stdout = Unpooled.buffer();
+        ByteBuf stderr = Unpooled.buffer();
+        docker.execStartCmd(execid).withDetach(false)
```
There's a lot of code in common with runCommand here. Doesn't need to be in this patch, but it should be factors out to something that takes a ```Consumer<byte[]>```

merlimat commented:
run cpp tests
